### PR TITLE
[SYCL][FPGA] Align clang with new spec of accessor_property_list

### PR DESCRIPTION
**Pull request by alexbatashev**
Signed-off-by: Mikhail Lychkov <mikhail.lychkov@intel.com>

### CODE REVIEW

File: clang/lib/Sema/SemaSYCL.cpp [hunk 1]
```diff
@@ -85,6 +85,10 @@ class Util {
   /// property_list class.
   static bool isPropertyListType(const QualType &Ty);
```
Comment (MrSidims) [inline clang/lib/Sema/SemaSYCL.cpp:86]: According to the new version of the spec (https://github.com/intel/llvm/pull/2439) `property_list` will no longer be used to specialize an accessor, hence no need to handle both `PropertyList` and `AccessorPropertyList`. This comment is applied to the code below as well.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 2]
```diff
@@ -3444,17 +3468,16 @@ bool Util::isSyclSpecConstantType(const QualType &Ty) {
...
   const StringRef &Name = "buffer_location";
-  std::array<DeclContextDesc, 4> Scopes = {
+  std::array<DeclContextDesc, 6> Scopes = {
       Util::DeclContextDesc{clang::Decl::Kind::Namespace, "cl"},
       Util::DeclContextDesc{clang::Decl::Kind::Namespace, "sycl"},
-      // TODO: this doesn't belong to property namespace, instead it shall be
-      // in its own namespace. Change it, when the actual implementation in SYCL
-      // headers is ready
+      Util::DeclContextDesc{clang::Decl::Kind::Namespace, "ext"},
+      Util::DeclContextDesc{clang::Decl::Kind::Namespace, "INTEL"},
```
Comment (MrSidims) [inline clang/lib/Sema/SemaSYCL.cpp:3480]: I find this nesting a little bit confusing - addressed the question in https://github.com/intel/llvm/pull/2439/files#r485446113

File: clang/lib/Sema/SemaSYCL.cpp [hunk 3]
```diff
@@ -1251,17 +1261,29 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
...
+
       TemplateArgument TA = TAL.get(0);
       const QualType TemplateArgTy = TA.getAsType();
-
-      if (TAL.size() > 5)
-        checkPropertyListType(TAL.get(5), Loc.getBegin());
       llvm::DenseSet<QualType> Visited;
       checkSYCLType(SemaRef, TemplateArgTy, Loc, Visited);
+
+      if (TAL.size() < 6) {
```
Comment (MrSidims) [inline clang/lib/Sema/SemaSYCL.cpp:1273]: Looks like we need to merge SYCL headers part first before enabling this diagnostics.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 4]
```diff
@@ -1402,19 +1424,21 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
...
 
   // Handle accessor properties. If any properties were found in
-  // the property_list - add the appropriate attributes to ParmVarDecl.
+  // the accessor_property_list - add the appropriate attributes to ParmVarDecl.
   void handleAccessorPropertyList(ParmVarDecl *Param,
                                   const CXXRecordDecl *RecordDecl,
                                   SourceLocation Loc) {
     const auto *AccTy = cast<ClassTemplateSpecializationDecl>(RecordDecl);
-    // TODO: when SYCL headers' part is ready - replace this 'if' with an error
     if (AccTy->getTemplateArgs().size() < 6)
```
Comment (MrSidims) [inline clang/lib/Sema/SemaSYCL.cpp:1432]: I don't think it's still required (we have the diagnostic).

File: clang/lib/Sema/SemaSYCL.cpp (hunk 3 above)
Comment (alexbatashev) [reply clang/lib/Sema/SemaSYCL.cpp]: I temporarily allowed old behavior, because to merge new headers we actually need this patch.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 5]
```diff
@@ -3472,14 +3460,15 @@ bool Util::isPropertyListType(const QualType &Ty) {
...
+  const StringRef &PropertyName = "buffer_location";
+  const StringRef &InstanceName = "instance";
   std::array<DeclContextDesc, 6> Scopes = {
-      Util::DeclContextDesc{clang::Decl::Kind::Namespace, "cl"},
-      Util::DeclContextDesc{clang::Decl::Kind::Namespace, "sycl"},
-      Util::DeclContextDesc{clang::Decl::Kind::Namespace, "ext"},
-      Util::DeclContextDesc{clang::Decl::Kind::Namespace, "INTEL"},
-      Util::DeclContextDesc{clang::Decl::Kind::Namespace, "property"},
-      Util::DeclContextDesc{Decl::Kind::ClassTemplateSpecialization, Name}};
+    Util::DeclContextDesc{Decl::Kind::Namespace, "cl"},
```
Comment (keryell) [inline clang/lib/Sema/SemaSYCL.cpp]: How does this work with SYCL 2020 `::sycl::` instead of `::cl::sycl::`?

Comment (alexbatashev) [reply clang/lib/Sema/SemaSYCL.cpp]: @keryell it should be fine. We define `cl` as `inline namespace`.

**Review by MrSidims — APPROVED** [review]

File: clang/lib/Sema/SemaSYCL.cpp [hunk 6]
```diff
@@ -1251,17 +1255,17 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
     assert(Util::isSyclAccessorType(Ty) &&
            "Should only be called on SYCL accessor types.");
 
-    const RecordDecl *RecD = Ty->getAsRecordDecl();
-    if (const ClassTemplateSpecializationDecl *CTSD =
-            dyn_cast<ClassTemplateSpecializationDecl>(RecD)) {
+    RecordDecl *RecD = Ty->getAsRecordDecl();
+    if (auto *CTSD = dyn_cast<ClassTemplateSpecializationDecl>(RecD)) {
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaSYCL.cpp:1259]: Why did you remove `const` quals from here?

File: clang/lib/Sema/SemaSYCL.cpp [hunk 7]
```diff
@@ -1251,17 +1255,17 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
...
       const TemplateArgumentList &TAL = CTSD->getTemplateArgs();
+
       TemplateArgument TA = TAL.get(0);
       const QualType TemplateArgTy = TA.getAsType();
+      llvm::DenseSet<QualType> Visited;
+      checkSYCLType(SemaRef, TemplateArgTy, Loc, Visited);
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaSYCL.cpp:1265]: What is the reason for this move?

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 8]
```diff
@@ -11033,9 +11033,10 @@ def err_sycl_compiletime_property_duplication : Error<
 def err_sycl_invalid_property_list_param_number : Error<
   "%0 must have exactly one template parameter">;
 def err_sycl_invalid_accessor_property_template_param : Error<
-  "Fifth template parameter of the accessor must be of a property_list type">;
-def err_sycl_invalid_property_list_template_param : Error<
-  "%select{property_list|property_list pack argument|buffer_location}0 "
+  "Sixth template parameter of the accessor must be of accessor_property_list "
```
Comment (Fznamznon) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:11036]: Clang doc says "Do not capitalize the diagnostic string"
```suggestion
  "sixth template parameter of the accessor must be of accessor_property_list "
```

**Review by MrSidims — APPROVED** [review]

File: clang/lib/Sema/SemaSYCL.cpp [hunk 9]
```diff
@@ -1251,15 +1255,17 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
     assert(Util::isSyclAccessorType(Ty) &&
            "Should only be called on SYCL accessor types.");
 
-    const RecordDecl *RecD = Ty->getAsRecordDecl();
+    RecordDecl *RecD = Ty->getAsRecordDecl();
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaSYCL.cpp:1258]: Still removed const qualifier. Why?

Comment (elizabethandrews) [inline clang/lib/Sema/SemaSYCL.cpp:1258]: Looks like qualifier was removed here as well. Is this intentional?

File: clang/lib/Sema/SemaSYCL.cpp [hunk 10]
```diff
@@ -1251,15 +1255,17 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
...
     if (const ClassTemplateSpecializationDecl *CTSD =
             dyn_cast<ClassTemplateSpecializationDecl>(RecD)) {
       const TemplateArgumentList &TAL = CTSD->getTemplateArgs();
+
```
Comment (elizabethandrews) [inline clang/lib/Sema/SemaSYCL.cpp:1262]: Extra lines here can also be removed. If checkAccessorType does not need to be modified in this patch, these changes will just be unrelated and random.

File: clang/test/CodeGenSYCL/accessor_inheritance.cpp [hunk 11]
```diff
@@ -67,13 +67,13 @@ int main() {
 // CHECK: [[ACC_FIELD:%[a-zA-Z0-9_]+]] = getelementptr inbounds %struct{{.*}}Base, %struct{{.*}}Base* [[BITCAST]], i32 0, i32 2
 // CHECK: [[ACC1_AS_CAST:%[a-zA-Z0-9_]+]] = addrspacecast %"class{{.*}}cl::sycl::accessor"* [[ACC_FIELD]] to %"class{{.*}}cl::sycl::accessor" addrspace(4)*
 // Default constructor call
-// CHECK: call spir_func void @_ZN2cl4sycl8accessorIcLi1ELNS0_6access4modeE1024ELNS2_6targetE2014ELNS2_11placeholderE0ENS0_13property_listIJEEEEC1Ev(%"class{{.*}}cl::sycl::accessor" addrspace(4)* [[ACC1_AS_CAST]])
+// CHECK: call spir_func void @_ZN2cl4sycl8accessorIcLi1ELNS0_6access4modeE1024ELNS2_6targetE2014ELNS2_11placeholderE0ENS0_6ONEAPI22accessor_property_listIJEEEEC1Ev(%"class._ZTSN2cl4sycl8accessorIcLi1ELNS0_6access4modeE1024ELNS2_6targetE2014ELNS2_11placeholderE0ENS0_6ONEAPI22accessor_property_listIJEEEEE.cl::sycl::accessor" addrspace(4)* [[ACC1_AS_CAST]])
```
Comment (elizabethandrews) [inline clang/test/CodeGenSYCL/accessor_inheritance.cpp:70]: I would prefer you kept {{.*}} in argument list. It makes the tests more readable

File: clang/test/SemaSYCL/buffer_location.cpp [hunk 12]
```diff
@@ -14,17 +14,17 @@ struct Base {
...
       AccField;
 };
 
 struct Captured : Base,
                   cl::sycl::accessor<char, 1, cl::sycl::access::mode::read,
                                      cl::sycl::access::target::global_buffer,
                                      cl::sycl::access::placeholder::false_t,
-                                     cl::sycl::property_list<
-                                         cl::sycl::property::buffer_location<1>>> {
+                                     cl::sycl::ONEAPI::accessor_property_list<
```
Comment (elizabethandrews) [inline clang/test/SemaSYCL/buffer_location.cpp:26]: Formatting looks weird

File: clang/test/CodeGenSYCL/accessor_inheritance.cpp [hunk 13]
```diff
@@ -67,13 +67,13 @@ int main() {
...
 // CHECK: [[BITCAST1:%[a-zA-Z0-9_]+]] = bitcast %struct{{.*}}Captured* [[GEP]] to i8*
-// CHECK: [[GEP1:%[a-zA-Z0-9_]+]] = getelementptr inbounds i8, i8* [[BITCAST1]], i64 24
+// CHECK: [[GEP1:%[a-zA-Z0-9_]+]] = getelementptr inbounds i8, i8* [[BITCAST1]], i64 20
 // CHECK: [[BITCAST2:%[a-zA-Z0-9_]+]] = bitcast i8* [[GEP1]] to %"class{{.*}}cl::sycl::accessor"*
 // CHECK: [[ACC2_AS_CAST:%[a-zA-Z0-9_]+]] = addrspacecast %"class{{.*}}cl::sycl::accessor"* [[BITCAST2]] to %"class{{.*}}cl::sycl::accessor" addrspace(4)*
 // Default constructor call
-// CHECK: call spir_func void @_ZN2cl4sycl8accessorIcLi1ELNS0_6access4modeE1024ELNS2_6targetE2014ELNS2_11placeholderE0ENS0_13property_listIJEEEEC2Ev(%"class{{.*}}cl::sycl::accessor" addrspace(4)* [[ACC2_AS_CAST]])
+// CHECK: call spir_func void @_ZN2cl4sycl8accessorIcLi1ELNS0_6access4modeE1024ELNS2_6targetE2014ELNS2_11placeholderE0ENS0_6ONEAPI22accessor_property_listIJEEEEC2Ev(%"class._ZTSN2cl4sycl8accessorIcLi1ELNS0_6access4modeE1024ELNS2_6targetE2014ELNS2_11placeholderE0ENS0_6ONEAPI22accessor_property_listIJEEEEE.cl::sycl::accessor" addrspace(4)* [[ACC2_AS_CAST]])
```
Comment (elizabethandrews) [inline clang/test/CodeGenSYCL/accessor_inheritance.cpp:76]: Here too.

File: clang/test/SemaSYCL/buffer_location.cpp (hunk 12 above)
Comment (alexbatashev) [reply clang/test/SemaSYCL/buffer_location.cpp]: Replaced property with an alias to make lines shorter. Not sure if there's another way to make it look less awkward.

**Review by Fznamznon — APPROVED** [review]

**Review by elizabethandrews — APPROVED** [review]
LGTM. Thanks!
